plugins/dispatch/src/main/java/org/qbicc/plugin/dispatch/DispatchTables.java

```diff
@@ -87,12 +87,17 @@ public VTableInfo getVTableInfo(LoadedTypeDefinition cls) {
 
     void buildFilteredVTable(LoadedTypeDefinition cls) {
         tlog.debugf("Building VTable for %s", cls.getDescriptor());
+        ReachabilityInfo reachabilityInfo = ReachabilityInfo.get(ctxt);
 
         ArrayList<MethodElement> vtableVector = new ArrayList<>();
         for (MethodElement m: cls.getInstanceMethods()) {
-            if (ctxt.wasEnqueued(m)) {
-                tlog.debugf("\tadding reachable method %s%s", m.getName(), m.getDescriptor().toString());
-                ctxt.registerAutoQueuedElement(m);
+            if (!m.isPrivate() && reachabilityInfo.isDispatchableMethod(m)) {
```

Comment:
I think this change to not put private methods in the vtable is correct, but I am not a totally reliable Java language lawyer... 
If it isn't correct, I'll need to have the invokeExact handling also make instance methods dispatchable....

Reply:
I agree that this seems correct.

Reply:
Sorry, late to the party on this but as of JDK 11 when NestMates feature was added, `invokevirtual` can now be used to invoke private methods.

Quoting from the JEP bug:
> With the change to the access rules, and with suitable adjustments to byte code rules, we can allow simplified rules for generating invocation bytecodes:
invokespecial for private nestmate constructors,
invokevirtual for private non-interface, nestmate instance methods,
invokeinterface for private interface, nestmate instance methods; and
invokestatic for private nestmate, static methods

https://bugs.openjdk.java.net/browse/JDK-8046171

Reply:
Thanks Dan.  I'll do a follow up PR to put the private methods back into vtables.  Skimming the JEP, I think we also need to put them into itables (which I think we also might not be doing currently).

Reply:
It's easiest to put them in the itables but I think we can avoid it.  With J9, we used the low bits of the itable index to indicate if this was a direct call or not so that the private methods didn't end up in the itable and were able to do the equivalent of a invokespecial after confirming the receiver implemented the interface.

Reply:
Given the closed world, we may be able to aggressively re-write `invokevirtual/invokeinterface` of private methods to `invokespecial` and keep them out of all vtable/itables.

Reply:
ok, so maybe we leave this as is for now and think further.

Reply:
I verified (empirically) that private methods are not using virtual dispatch even when made accessible via nestmates despite using the `invokevirtual` bytecode.

```java
class Scratch {
    public static void main(String[] args) {
        final Sub sub = new Sub();
        System.out.println(((Base) sub).doIt());
    }

    private static class Base {
        private int doIt() {
            return 123;
        }
    }

    private static class Sub extends Base {
        private int doIt() {
            return 456;
        }
    }
}
```

Output:
```text
123

Process finished with exit code 0
```

Bytecodes:
```javap
Classfile /Users/david/Library/Caches/JetBrains/IdeaIC2021.2/compile-server/qbicc-parent_dc4ec81f/scratches/out/Scratch.class
[...]
  public static void main(java.lang.String[]);
    descriptor: ([Ljava/lang/String;)V
    flags: (0x0009) ACC_PUBLIC, ACC_STATIC
    Code:
      stack=2, locals=2, args_size=1
         0: new           #7                  // class Scratch$Sub
         3: dup
         4: invokespecial #9                  // Method Scratch$Sub."<init>":()V
         7: astore_1
         8: getstatic     #10                 // Field java/lang/System.out:Ljava/io/PrintStream;
        11: aload_1
        12: invokevirtual #16                 // Method Scratch$Base.doIt:()I
        15: invokevirtual #22                 // Method java/io/PrintStream.println:(I)V
        18: return
      LineNumberTable:
        line 3: 0
        line 4: 8
        line 5: 18
      LocalVariableTable:
        Start  Length  Slot  Name   Signature
            0      19     0  args   [Ljava/lang/String;
            8      11     1   sub   LScratch$Sub;
}
SourceFile: "scratch_1.java"
NestMembers:
  Scratch$Sub
  Scratch$Base
InnerClasses:
  private static #45= #7 of #28;          // Sub=class Scratch$Sub of class Scratch
  private static #46= #17 of #28;         // Base=class Scratch$Base of class Scratch

Classfile /Users/david/Library/Caches/JetBrains/IdeaIC2021.2/compile-server/qbicc-parent_dc4ec81f/scratches/out/Scratch$Base.class
[...]
  private int doIt();
    descriptor: ()I
    flags: (0x0002) ACC_PRIVATE
    Code:
      stack=1, locals=1, args_size=1
         0: bipush        123
         2: ireturn
      LineNumberTable:
        line 9: 0
      LocalVariableTable:
        Start  Length  Slot  Name   Signature
            0       3     0  this   LScratch$Base;
}
SourceFile: "scratch_1.java"
NestHost: class Scratch
InnerClasses:
  private static #22= #7 of #19;          // Base=class Scratch$Base of class Scratch

Classfile /Users/david/Library/Caches/JetBrains/IdeaIC2021.2/compile-server/qbicc-parent_dc4ec81f/scratches/out/Scratch$Sub.class
[...]
  private int doIt();
    descriptor: ()I
    flags: (0x0002) ACC_PRIVATE
    Code:
      stack=1, locals=1, args_size=1
         0: sipush        456
         3: ireturn
      LineNumberTable:
        line 15: 0
      LocalVariableTable:
        Start  Length  Slot  Name   Signature
            0       4     0  this   LScratch$Sub;
}
SourceFile: "scratch_1.java"
NestHost: class Scratch
InnerClasses:
  private static #22= #2 of #19;          // Base=class Scratch$Base of class Scratch
  private static #23= #7 of #19;          // Sub=class Scratch$Sub of class Scratch
```

So our present policy of converting invocations on private methods to exact invocations - and consequently keeping them out of vtables - still seems correct to me despite the usage of the `invokevirtual` bytecode.